### PR TITLE
N°2651 - Removal of "Test" dirs within dependencies handled using composer

### PR DESCRIPTION
**Pull request by bruno-ds**
Note: 

since it was the problem with the previous solution, the `iTopComposer::ListAllowedTestDir()`
MUST be reread, for now it only contains two entries : 

			APPROOT.'lib/twig/twig/src/Node/Expression/Test/',
			APPROOT.'lib/twig/twig/lib/Twig/Node/Expression/Test/',

### CODE REVIEW

File: sources/application/composer/iTopComposer.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+			APPROOT.'lib/symfony/http-foundation/Tests/',
+			APPROOT.'lib/symfony/http-kernel/Tests/',
+			APPROOT.'lib/symfony/routing/Tests/',
+			APPROOT.'lib/symfony/stopwatch/Tests/',
+			APPROOT.'lib/symfony/twig-bridge/Tests/',
+			APPROOT.'lib/symfony/twig-bundle/Tests/',
+			APPROOT.'lib/symfony/var-dumper/Tests/',
+			APPROOT.'lib/symfony/web-profiler-bundle/Tests/',
+			APPROOT.'lib/symfony/yaml/Tests/',
+			APPROOT.'lib/symfony/debug/Resources/ext/tests/',
```
Comment (piRGoif) [inline sources/application/composer/iTopComposer.php]: For the following :

'lib/symfony/debug/Resources/ext/tests/'
'lib/symfony/framework-bundle/Tests/Fixtures/TestBundle/FooBundle/Controller/Test/'

Might not be a good idea to remove... are we sure they really contains tests only ?

Comment (bruno-ds) [reply sources/application/composer/iTopComposer.php]: after analysing this : 
 - `lib/symfony/debug/Resources/ext` is C code (a php extention), so it is out of scope and the test dir can be excluded.

 - `lib/symfony/framework-bundle/Tests` is a test dir, so it subdir can be excluded

**Review by piRGoif — DISMISSED** [review]

**Comment by Molkobain** [comment]
Seems all good to me. Just wondering about the location of the iTopComposer.php file. Is /sources/application the best place? This folder is meant mostly for the GUI of the admin. console. Also, shouldn't we put a namespace to it if it's in the /sources folder like all the other files ?

**Comment by bruno-ds** [comment]
I admit that I struggled to find where to put it, and wasn't happy with my choice.
Even more now that you point the GUI related semantic of this for.
Feel free to move it :)

For the namespace, has said many sages during hard time: whoops!

**Review by Molkobain — DISMISSED** [review]

**Comment by piRGoif** [comment]
> Just wondering about the location of the iTopComposer.php file. Is /sources/application the best place?

I agree too !
Why not simply /lib or /lib/bin ?

**Comment by piRGoif** [comment]
> 
> > Just wondering about the location of the iTopComposer.php file. Is /sources/application the best place?
>
> Why not simply /lib or /lib/bin ?

Discussed with Bruno : not a good idea, as this dir content is handled by Composer !
Also, the goal was to have a specific namespace in iTopComposer.php, and so putting it in /application made it working natively with our autoloader.

**Comment by piRGoif** [comment]
@bruno-ds I've juste pushed (5187834) a modification for the \iTopComposer::ListAllTestDir method : use Spl classes instead of multiple glob patterns.
Results are the same now (I fixed some problems on windows with DIRECTORY_SEPARATOR)

File: sources/application/composer/iTopComposer.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+	public function ListAllTestDir()
+	{
+		$aAllTestDirs = array();
+		$sPath = realpath(APPROOT.'lib');
+
+		$oDirectoryIterator = new RecursiveDirectoryIterator($sPath);
+		$iterator = new RecursiveIteratorIterator(
+			$oDirectoryIterator,
+			RecursiveIteratorIterator::CHILD_FIRST);
```
Comment (bruno-ds) [inline sources/application/composer/iTopComposer.php]: https://www.php.net/manual/fr/class.filesystemiterator.php#filesystemiterator.constants

I propose : 
FilesystemIterator::SKIP_DOTS
FilesystemIterator::UNIX_PATHS

Comment (piRGoif) [reply sources/application/composer/iTopComposer.php]: Yeah, seems perfect ! Can you add it as a code suggestion ? (https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request see point 6)

**Comment by Molkobain** [comment]
I'm still troubled by this file being in the /sources/application folder as it is really not part of the admin. console GUI... 😕

As this is more of an internal tool or something, would a /bin or /.make be suitable?

**Comment by piRGoif** [comment]
> As this is more of an internal tool or something, would a /bin or /.make be suitable?

Indeed, I would recommend renaming .make to .bin, and moving iTopComposer.php file there ?

**Comment by odain-cbd** [comment]
Pull request already merged manually.

**Comment by piRGoif** [comment]
Updates tags & milestone :)
Are you ok to delete branch ?

**Comment by odain-cbd** [comment]
Thanks for the reminder. we planned to do it with Bruno and forgot...

**Pull request closed without merging**